spec/std/wait_group_spec.cr

```diff
@@ -0,0 +1,104 @@
+require "spec"
+require "wait_group"
+
+describe WaitGroup do
+  describe "add" do
+    it "can't decrement to a negative counter" do
+      wg = WaitGroup.new
+      wg.add(5)
+      wg.add(-3)
+      expect_raises(Exception) { wg.add(-5) }
+    end
+  end
+
+  describe "done" do
+    it "can't decrement to a negative counter" do
+      wg = WaitGroup.new
+      wg.add(1)
+      wg.done
+      expect_raises(Exception) { wg.done }
+    end
+  end
+
+  it "waits until concurrent executions are finished" do
+    wg1 = WaitGroup.new
+    wg2 = WaitGroup.new
+
+    8.times do
+      wg1.add(16)
+      wg2.add(16)
+      exited = Channel(Bool).new(16)
+
+      16.times do
+        spawn do
+          wg1.done
+          wg2.wait
+          exited.send(true)
+        end
+      end
+
+      wg1.wait
+
+      16.times do
+        select
+        when exited.receive
+          fail "WaitGroup released group too soon"
+        else
+        end
+        wg2.done
+      end
```

Comment:
**thought:** I was wondering if the receive action really needs to run 16 times. Even a single receive would fail the spec.
So I *think* this could technically be unrolled to `15.times { wg2.done }; select ...; wg2.done`. Not sure if it improves much, but it might have a slight chance to describe the purpose a bit more clearly?

Reply:
To be honest I'm mostly following Go's test (which is quite smart). I read it as:

- I musn't receive a message (didn't even start to decrement);
- I decrement by one;
- I still musn't receive a message;
- ...
- I decrement by one (counter reaches zero)
- I must receive all messages.

I think this behavior sounds more realistic and yields fibers between each call to `#done` which may help trigger a potential bug, otherwise we'd just merely decrement a counter in a loop.